### PR TITLE
feat: buy token changes dynamically according to token page

### PR DESCRIPTION
**Pull request by evandrosaturnino**
## Description

1- Whenever the user opens a token page (After the token get's updated as well), it dispatches an action SET_TOKEN to populate a global store in redux with the current opened asset
2- Added a conditional validation in the useTradeRoutes to check if the current asset is valid (not undefined or null), if it is in 'ethereum' chain, and ultimately if it's not ETH itself to avoid ETH <> ETH situation
3- Implemented/adjusted the useEffect functions to make the application be tightly bound, to work accordingly only when there are expected updates

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [x] :nail_care: New Feature (Non-breaking Change)

## Issue (if applicable)

closes issue #934

## Testing

1. Run the application 
2. Connect the Wallet
3. Pick any token in the tokens tab to see the trading asset changing accordingly

## Screenshot

https://user-images.githubusercontent.com/60587527/153696242-bfa058f8-a4e8-4ed4-b0fb-3213311f3135.mp4

### CODE REVIEW

File: src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts [hunk 1]
```diff
@@ -18,12 +19,16 @@ export const useTradeRoutes = (): {
   const buyAsset = getValues('buyAsset')
   const sellAsset = getValues('sellAsset')
   const assets = useSelector(selectAssets)
+  const asset = assetStore.getState()
```
Comment (0xdef1cafe) [inline src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts:22]: there's a much easier way to get the current asset id - take a look here https://github.com/shapeshift/web/blob/develop/src/pages/Assets/Asset.tsx#L57

File: src/state/slices/assetsSlice/assetStore.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+export function feedAssetStore(asset: Asset) {
+  return {
+    type: 'SET_ASSET',
+    asset
+  }
+}
+
+export function reducer(state = initialAsset, action: any) {
```
Comment (0xdef1cafe) [inline src/state/slices/assetsSlice/assetStore.ts:32]: take a look at the comment above about how to get the current asset id - we won't need a new global state slice just to manage this :)

**Review by 0xdef1cafe — CHANGES_REQUESTED** [review]

File: src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts (hunk 1 above)
Comment (evandrosaturnino) [reply src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts:22]: Thanks for reviewing, in fact, this was my first attempt, I called useAsset from useTradeRoutes at first, but it returned an error due to the fact that both Asset.tsx params { chain, tokenId } were coming undefined, so I thought of creating a redux store instead of refactoring an already functional and important hook
image

**Comment by evandrosaturnino** [comment]
> @evandrosaturnino - are you able to incorporate @0xdef1cafe feedback and get an updated PR up for this? I would trust that @0xdef1cafe is probably correct that a simpler solution could be achieved here.

@0xean @0xdef1cafe Absolutely, in order for me to be able to call the useAsset hook directly from useTradeRoutes, was necessary to do a small refactoring on the hook, precisely in the way how the params were being returned with useParams() react-router hook.
The problem was that since the useAsset hook was being called from `useTradeRoutes`, the `useParams` was returning the params of the pathname `/trade/input` from `useTradeRoutes`. That's why `AssetId` and `Chain` were being returned undefined.
So following the same flow found in the application in other hooks and components, I used the matchPath hook to match the correct pathname in order to bring the correct params.

**Comment by evandrosaturnino** [comment]
> @evandrosaturnino - are you able to incorporate @0xdef1cafe feedback and get an updated PR up for this? I would trust that @0xdef1cafe is probably correct that a simpler solution could be achieved here.

@0xean @0xdef1cafe Absolutely, in order for me to be able to call the useAsset hook directly from useTradeRoutes, was necessary to do a small refactoring on the hook, precisely in the way how the params were being returned with useParams() react-router hook.
The problem was that since the useAsset hook was being called from useTradeRoutes, the useParams was returning the params of the pathname /trade/input from useTradeRoutes. That's why AssetId and Chain were being returned undefined.
So following the same flow found in the application in other hooks and components, I used the matchPath hook to match the correct pathname in order to bring the correct params.

**Comment by evandrosaturnino** [comment]
Added useAsset hook fn call in useTradeRoutes.test as well, unit tests are passing as expected.
commit: c69091a43c25da1a4096b5f6398efc1f14513018
Behaviour flow working as expected: 

https://user-images.githubusercontent.com/60587527/154822405-d95ed01c-422b-409d-bdbf-284eeb6194ce.mp4

**Review by gomesalexandre — CHANGES_REQUESTED** [review]
A few minor changes requested in comments, namely about typings.

One more thing I'm thinking: `useBrowserRouter` is a context made for convenience in /defi section that memoizes `useLocation()`, `useHistory()`, `useParams()` and `useQuery()` calls and returns them as an object.

This memoization allows us not to call `useLocation()` again on `TradeRoutes` routing but we should do a regression testing of the asset section to see if nothing breaks, since this usage of `react-router-dom` methods is non-standard.

File: src/pages/Assets/Asset.tsx [hunk 3]
```diff
@@ -42,16 +51,16 @@ export const useAsset = () => {
 export const Asset = () => {
   const { asset, marketData } = useAsset()
 
-  return !(asset && marketData) ? (
+  return asset && marketData ? (
     <Page style={{ flex: 1 }} key={asset?.tokenId}>
       <Flex role='main' flex={1} height='100%'>
-        <LoadingAsset />
+        <AssetAccountDetails assetId={asset.caip19} />
```
[the remark this answers is not in the document]
Comment (evandrosaturnino) [reply src/pages/Assets/Asset.tsx]: Thanks for reviewing it, resolved in commit ba7488ec2ec1889a2cf4b3f1000b2dde7c1be928

File: src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts [hunk 4]
```diff
@@ -18,12 +19,16 @@ export const useTradeRoutes = (): {
...
   const assets = useSelector(selectAssets)
+  const { asset } = useAsset()
 
   const setDefaultAssets = useCallback(async () => {
     try {
       const [sellAssetId, buyAssetId] = getDefaultPair()
       const sellAsset = assets[sellAssetId]
-      const buyAsset = assets[buyAssetId]
+      const buyAsset =
+        asset?.chain === 'ethereum' && asset?.caip19 !== sellAssetId
```
[the remark this answers is not in the document]
Comment (evandrosaturnino) [reply src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts:31]: Thanks for reviewing it, resolved in commit 396ba769a64ae229eaf9e3191d4b28ca414f32c9

**Comment by 0xdef1cafe** [comment]
> A few minor changes requested in comments, namely about typings.
> 
> One more thing I'm thinking: `useBrowserRouter` is a context made for convenience in /defi section that memoizes `useLocation()`, `useHistory()`, `useParams()` and `useQuery()` calls and returns them as an object.
> 
> In this Pr implementation, this memoization allows us not to call `useLocation()` again on `TradeRoutes` routing and keep the `/assets/` route as a location but we should do a regression testing of the asset section to see if nothing breaks, since this usage of `react-router-dom` methods is non-standard.

hey @evandrosaturnino - i agree with @gomesalexandre here.

i tested the account asset pages, i.e. click on accounts, then a token (e.g. usdc) under the ethereum account, and the default trade pair is still showing eth -> fox, rather than eth -> usdc.

can you please take a look at this case?

**Comment by evandrosaturnino** [comment]
> > A few minor changes requested in comments, namely about typings.
> > One more thing I'm thinking: `useBrowserRouter` is a context made for convenience in /defi section that memoizes `useLocation()`, `useHistory()`, `useParams()` and `useQuery()` calls and returns them as an object.
> > In this Pr implementation, this memoization allows us not to call `useLocation()` again on `TradeRoutes` routing and keep the `/assets/` route as a location but we should do a regression testing of the asset section to see if nothing breaks, since this usage of `react-router-dom` methods is non-standard.
> 
> hey @evandrosaturnino - i agree with @gomesalexandre here.
> 
> i tested the account asset pages, i.e. click on accounts, then a token (e.g. usdc) under the ethereum account, and the default trade pair is still showing eth -> fox, rather than eth -> usdc.
> 
> can you please take a look at this case?

Great! Thank you for reviewing it.
I refactored the code: Since I noticed that the Trade component already had props with assetId and chainId, I thought it would be easier to just pass them as params, following what was done in the accounts-related components.
Commit: https://github.com/shapeshift/web/pull/1008/commits/3da881fa6ee6d65edd914cb7596ae34b88e81449

About the problem with the account asset, I found out that the problem was in fact two different things: First, the props in the account path were structured differently, instead of assetId and chainId, it was accountID and subAssetId, so I validated this difference, and the second problem was that the props and path in the accounts are decoded. So I followed the same solution already used before in the application before, an example can be seen here: https://github.com/shapeshift/web/blob/5f28424957b679db0ff9412dc25c8d1cf0483500/src/pages/Accounts/Account.tsx#L17

Right now the functionality is working as expected, with a simpler integration and implementation. Therefore I think is ready for another round.
@0xdef1cafe

**Comment by gomesalexandre** [comment]
> instead of assetId and chainId, it was accountID and assetSubId, so I validated this difference,

Have a look at https://github.com/shapeshift/web/pull/1033/files#diff-fb47c0e7f76bb866d51fa66f6df489352e66678401202e0c3516aa070d0d0a9fR24 , it will make it more clear why the route params are this way now.
tl;dr is previously, `assetId` was considered a single route param, but now, we parse it as `<chainId>/<assetSubId>`. This is to parse the left and right side of the / separtor of CAIP19.

**Review by gomesalexandre — APPROVED** [review]
Great job @evandrosaturnino! A few minor comments which are not blocking.

Tested from `/dashboard`, `/accounts` and `/assets` route. The flow previously mentioned by @0xdef1cafe is now fixed and the right asset is always shown as a buy asset.

**Comment by evandrosaturnino** [comment]
> Great job @evandrosaturnino! A few minor comments which are not blocking.
> 
> Tested from `/dashboard`, `/accounts` and `/assets` route. The flow previously mentioned by @0xdef1cafe is now fixed and the right asset is always shown as a buy asset.

Thanks for the review! The suggestions were committed!

**Review by gomesalexandre — APPROVED** [review]

File: src/components/Trade/Trade.tsx [hunk 5]
```diff
@@ -36,12 +44,17 @@ export const Trade = () => {
       fiatAmount: undefined
     }
   })
-
+  const match = useParams<MatchParams>()
+  const assetId = !match.accountId
+    ? `${match?.chainId}/${match?.assetSubId}`
+    : `${decodeURIComponent(match?.assetId)}`
```
Comment (cjthompson) [inline src/components/Trade/Trade.tsx:50]: This feels really brittle to me. If we change how to route to asset pages, that will break this and it's not going to be obvious to a developer that this relies on a specific route format.

File: src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts [hunk 6]
```diff
@@ -19,12 +22,18 @@ export const useTradeRoutes = (): {
...
     try {
       const [sellAssetId, buyAssetId] = getDefaultPair()
       const sellAsset = assets[sellAssetId]
-      const buyAsset = assets[buyAssetId]
+      // TODO: this is fine for now, since we can only trade ETH/ERC20
+      // but should be refactored to adequate incoming new chains - i.e. thorchain swaps
+      const buyAsset =
+        asset?.chain === 'ethereum' && asset?.caip19 !== sellAssetId
+          ? assets[asset.caip19]
+          : assets[buyAssetId]
```
Comment (cjthompson) [inline src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts:36]: Why do we need chain specific code here? The `caip19` aka `assetId` string is sufficient to identify a specific asset.

**Comment by cjthompson** [comment]
I think there is a much simpler way to implement this.

In `AssetAccountDetails`, the `<TradeCard />` component is rendered. Just pass in props with 

```ts
props: { defaultBuyAssetId: CAIP19, defaultSellAssetId: CAIP19 }
```

Then pass in the caip19 values for the suggested buy and sell asset.

**Comment by 0xdef1cafe** [comment]
hey @evandrosaturnino - i agree with @cjthompson comments here regarding fragility and simplicity of implementation.
